Add "request access" experimental feature

erzhtor opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/47445.

[Demo video](https://www.loom.com/share/8c9dd4d916f8421b96bb9be35dad0960)

## Screenshots

| Description | Screenshot |
| --: | :-- |
| Sign in page when builtin auth provider allowSignup=false | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219633483-0bd31e33-2afc-47e7-9727-940b946f52db.png">  |
| Request access form page | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219635313-8a086977-d250-4e81-9e1c-6d0db43e20de.png"> |
| Request access error submission page: user with email already exists | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219635840-ec636ecf-8b16-463a-b78b-0d9a511e8ed7.png"> |
| Request access error submission page: access request with email already exists | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219638764-55eb77d6-6b09-44bb-8af2-9b54acfd0270.png"> |
| Request access submission success page | <img width="1724" alt="image" src="https://user-images.githubusercontent.com/6717049/219636963-62c2b44d-2c37-469a-b8eb-3b1afc5205e0.png"> |
| Site admin global navbar notification | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896058-3805ecbf-ff79-4e6e-8ad2-b4a8b3ae0cbd.png"> |
| Site admin / Access Requests page | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896187-bec260e1-4b92-44a7-96c3-b29ad507a434.png"> |
| Site admin / Access Requests / Approve | <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896585-961dfba0-15b7-4e9e-9124-6d313d59c1c8.png"> |
| Site admin / Access Requests / Approve with error |  <img width="1728" alt="image" src="https://user-images.githubusercontent.com/6717049/220896826-a2263dd7-bee3-4ae9-ab10-759ca7de8d40.png"> |

## Test plan
- `sg start`
- Open http://localhost:3080/site-admin/configuration and disable builtin auth provider allowSignup
- Open http://localhost:3080/sign-in in incognito
- Click the "Request access" link
- Fill the form and submit a request
- Go to http://localhost:3080 (as site admin)
- Check there is an "Access requests #" button in the global navbar
- Click to that button
- Approve/Reject access request

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

erzhtor commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,120 @@
+		)
+		switch {
+		case database.IsAccessRequestUserWithEmailExists(err):
+			// TODO: clarify how to handle this error, since this will be shown to unauthenticated user
+			message = "A user with this email already exists."
+			statusCode = http.StatusConflict
```
TODO: clarify whether to show this message to unauthenticated user or not?

evict replied:
Maybe we can do something similar as we [do in the password reset form](
https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/internal/auth/userpasswd/reset_password.go?L69-77. )?

erzhtor replied:
Thank you! Yes, that sounds the right approach! 👍

erzhtor commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,120 @@
+			message = "A user with this email already exists."
+			statusCode = http.StatusConflict
+		case database.IsAccessRequestWithEmailExists(err):
+			// TODO: clarify how to handle this error, since this will be shown to unauthenticated user
+			message = "A access request was already created previously."
+			statusCode = http.StatusConflict
```
TODO: clarify whether to show this message to unauthenticated user or not?

erzhtor replied:
Used similar to https://github.com/sourcegraph/sourcegraph/pull/47741#discussion_r1109992482 approach

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.11% (+3.14 kb) | 0.14% (+19.92 kb) 🔺 | 0.14% (+16.78 kb) 🔺 | 0.26% (+2) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/erzhtor/add-request-access-feature/compare-report.html;tab=live_object) for a full comparison between the commits cba438a and 3cb0cc6 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -9362,3 +9362,111 @@ extend type Mutation {
+"""
+A list of access requests
+"""
+type AccessRequests {
```
I think we should stay consistent in our GraphQL schema:

```suggestion
type AccessRequestConnection {
```

See the others: https://sourcegraph.sourcegraph.com/search?q=context%3Asourcegraph+Connection+%7B&groupBy=repo

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -9362,3 +9362,111 @@ extend type Mutation {
+    A list of access requests
+    """
+    nodes(
+        """
+        Returns the "limit" number access_requests from the list.
+        """
+        limit: Int
+        """
+        Skips initial "offset" number of access_requests.
+        """
+        offset: Int
+        """
+        Returns access_requests ordered by a given column.
+        """
+        orderBy: AccessRequestOrderBy
+        """
+        Returns ordered access_requests in descending order provided by orderBy field.
+        """
+        descending: Boolean
```
I'll let @thenamankumar comment but I think we have different standards now for connections. Might be wrong though.

erzhtor replied:
https://github.com/sourcegraph/sourcegraph/pull/47741#discussion_r1115552071

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -9362,3 +9362,111 @@ extend type Mutation {
+    status: AccessRequestStatus!
+}
+
+extend type Query {
+    """
+    List access requests.
+    """
+    accessRequests(status: AccessRequestStatus): AccessRequests!
+}
+
+extend type Mutation {
```
Why add this down there and not up top with the other Query and Mutations?

erzhtor replied:
To collocate changes.

mrnugget commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+func handleEnabledCheck(logger log.Logger, w http.ResponseWriter) (handled bool) {
+	experimentalFeatures := conf.Get().ExperimentalFeatures
+	if experimentalFeatures != nil && experimentalFeatures.AccessRequests != nil && !experimentalFeatures.AccessRequests.Enabled {
+		logger.Error("Request access is disabled.")
+		http.Error(w, "Request access is disabled.", http.StatusConflict)
```
```suggestion
		logger.Error("experimental feature accessRequests is disabled, but received request")
		http.Error(w, "experimental feature accessRequests is disabled, but received request", http.StatusForbidden)
```

mrnugget commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+	}
+}
+
+// HandleRequestAccessEnabledCheck checks whether request access experimental feature is explicitly disabled
+func handleEnabledCheck(logger log.Logger, w http.ResponseWriter) (handled bool) {
```
I'd rename this and turn the boolen around:

```suggestion
func accessRequestsFeatureEnabled(logger log.Logger, w http.ResponseWriter) (handled bool) {
```

Easier to understand. `handle...` sounds like it has side-effects and checking whether a feature is explicitly disabled is a bit weird for an experimental feature that needs to be enabled.

pjlast replied:
```suggestion
// handleEnabledCheck checks whether request access experimental feature is explicitly disabled
func handleEnabledCheck(logger log.Logger, w http.ResponseWriter) (handled bool) {
```

mrnugget commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+		return
+	}
+
+	const defaultErrorMessage = "Request access failed unexpectedly."
```
I'd move this down to its usage. It's ~20lines removed from that and it doesn't make a difference whether it's a `const` here or a string literal down there.

mrnugget commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+	const defaultErrorMessage = "Request access failed unexpectedly."
+
+	if err := userpasswd.CheckEmailFormat(data.Email); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
I think status `422 Unprocessable Entity` is more correct.

thenamankumar commented:
Few questions:
- How do we notify the users that they have been provided access? Do we send an email, or the site admin needs to notify them?
- Are we planning to integrate this with the in-progress RBAC system? 
- Can we use an icon in the top nav bar rather than "Access Requests" full text?

mrnugget commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+			// TODO: clarify how to handle this error, since this will be shown to unauthenticated user
+			message = "A access request was already created previously."
+			statusCode = http.StatusConflict
+		case errcode.PresentationMessage(err) != "":
```
```suggestion
		case message := errcode.PresentationMessage(err); message != "":
```

Hmm, not sure if it works, but maybe it does?

erzhtor replied:
Doesn't seem to work :(, does not understand syntax.

thenamankumar commented on client/web/src/Layout.tsx:
```diff
@@ -114,10 +114,11 @@ export const Layout: React.FC<LegacyLayoutProps> = props => {
-        location.pathname === PageRoutes.SignUp ||
-        location.pathname === PageRoutes.PasswordReset ||
-        location.pathname === PageRoutes.Welcome
+        routeMatch === PageRoutes.SignIn ||
```
```javascript
[
  PageRoutes.SignIn,
  PageRoutes.SignUp,
  ...
].includes(routeMatch)
```

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+func (s *accessRequestStore) Create(ctx context.Context, accessRequest *types.AccessRequest) (*types.AccessRequest, error) {
+	// We don't allow adding a new request_access with an email address that has already been
+	// verified by another user.
+	exists, _, err := basestore.ScanFirstBool(s.Query(ctx, sqlf.Sprintf("SELECT TRUE WHERE EXISTS (SELECT FROM user_emails WHERE email = %s AND verified_at IS NOT NULL)", accessRequest.Email)))
```
```suggestion
	exists, _, err := basestore.ScanFirstBool(s.Query(ctx, sqlf.Sprintf("SELECT TRUE FROM user_emails WHERE email = %s AND verified_at IS NOT NULL", accessRequest.Email)))
```

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	// We don't allow adding a new request_access with an email address that has already been used
+	exists, _, err = basestore.ScanFirstBool(s.Query(ctx, sqlf.Sprintf("SELECT TRUE WHERE EXISTS (SELECT FROM access_requests WHERE email = %s)", accessRequest.Email)))
```
```suggestion
	exists, _, err = basestore.ScanFirstBool(s.Query(ctx, sqlf.Sprintf("SELECT TRUE FROM access_requests WHERE email = %s)", accessRequest.Email)))
```

malomarrec commented:
Great stuff 🔥

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+	)
+	var data types.AccessRequest
+
+	if err := s.QueryRow(ctx, q).Scan(&data.ID, &data.CreatedAt, &data.UpdatedAt, &data.Name, &data.Email, &data.Status, &data.AdditionalInfo); err != nil {
+		return nil, errors.Wrap(err, "scanning access_request")
+	}
+
+	return &data, nil
+}
+
+func (s *accessRequestStore) GetByID(ctx context.Context, id int32) (*types.AccessRequest, error) {
+	row := s.QueryRow(ctx, sqlf.Sprintf("SELECT id, created_at, updated_at, name, email, status, additional_info FROM access_requests WHERE id = %s", id))
+	var node types.AccessRequest
+
+	if err := row.Scan(&node.ID, &node.CreatedAt, &node.UpdatedAt, &node.Name, &node.Email, &node.Status, &node.AdditionalInfo); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, &errAccessRequestNotFound{ID: id}
+		}
+		return nil, err
+	}
+
+	return &node, nil
+}
+
+func (s *accessRequestStore) Update(ctx context.Context, accessRequest *types.AccessRequest) (*types.AccessRequest, error) {
+	q := sqlf.Sprintf(accessRequestUpdateQuery, accessRequest.Status, accessRequest.ID)
+	var updated types.AccessRequest
+
+	if err := s.QueryRow(ctx, q).Scan(&updated.ID, &updated.CreatedAt, &updated.UpdatedAt, &updated.Name, &updated.Email, &updated.Status, &updated.AdditionalInfo); err != nil {
```
Highly recommend using a `scanAccessRequest` function here. See the `scanHorse` thing here: https://docs.sourcegraph.com/dev/background-information/basestore#using-basestore-helpers-for-scanning

erzhtor replied:
TIL, thank you, updated! 👍

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+
+func (s *accessRequestStore) Count(ctx context.Context, opt *AccessRequestsFilterOptions) (int, error) {
+	q := sqlf.Sprintf("SELECT COUNT(*) FROM access_requests WHERE (%s)", sqlf.Join(opt.sqlConditions(), ") AND ("))
+	var count int
+	if err := s.QueryRow(ctx, q).Scan(&count); err != nil {
+		return 0, err
+	}
+	return count, nil
```
```suggestion
	return basestore.ScanInt(s.QueryRow(ctx, q))
```

I _think_ this should work

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+	defer rows.Close()
+
+	nodes := make([]*types.AccessRequest, 0)
+	for rows.Next() {
```
See above, if you create a `scanAccessRequest` function you can create a new `basestore.NewSliceScanner` and use that here.

erzhtor replied:
TIL, thank you, updated! 👍

mrnugget commented on internal/database/access_requests_test.go:
```diff
@@ -0,0 +1,261 @@
+		assert.NoError(t, err)
+		assert.Equal(t, len(accessRequests), 1)
+	})
+}
```
Good tests!

mrnugget commented on migrations/frontend/1675850599_add_access_requests_table/up.sql:
```diff
@@ -0,0 +1,19 @@
+    name TEXT NOT NULL,
+    email TEXT NOT NULL UNIQUE,
+    additional_info TEXT,
+    status access_request_status NOT NULL DEFAULT 'PENDING'
```
Personal opinion: I'd just use a `text` here.

erzhtor replied:
Can you elaborate why? I thought having a custom type is better to keep data aligned and consistent.

mrnugget replied:
Do we plan on modifying data directly via SQL? Or from another application? I don't think we do, so all changes will go through our database layer in Go, right? So that's already a lot of advantages of enums that go out of the window. 

And if we, some day, want to introduce a third rate (`declined`, `re-reqested`, ... whatever), we need to write a separate database migration to re-create the type and migrate the old data. We can't just add a new string in our Go code.

So I think the benefits in our use case (single writer, versioned application code with migrations) are minimal compared to the hassle of having to migrate a table if the enum will change in the future.

erzhtor replied:
Thank you for the detailed explanation! 👍  Makes sense, updated 4f07261271dd1545743e31c93ba58a0f06cf2908.

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	return &data, nil
+}
```
You have 3 queries in this method without a transaction. I'm pretty sure you need a transaction here. But also: why check for this in here? this is business logic I'd expect in a `backend` service (example: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/backend/own.go?subtree=true) but not in the store. I'm not hard blocking on this, but if I had to write this I'd write 3 separate store methods and then use a transaction in a service and provide meaningful data there. Example: batch changes business logic in a "service" to make sure no duplicates are created: https://github.com/sourcegraph/sourcegraph/blob/98679e18623440f1994e3a7e37bc2c6735b27a44/enterprise/internal/batches/service/service.go#L200-L219

That's *not* in the store.

erzhtor replied:
I agree with running logic in a transaction; thanks for the suggestion! 

> this is business logic I'd expect in a backend service

Regarding moving into a separate service, I was following a similar to the existing [database/users.go / Create](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5bd36a45c113f87d15948f3e6bc174e96fe18f36/-/blob/internal/database/users.go?L257), which has all such constraint checks inside store.Create. I thought, it makes sense since these constraints should not be omitted in any cases, and thus be part of the store. WDYT?

mrnugget replied:
`Users.Create` is some of the oldest code we have in the codebase, possibly not the best thing to look towards for guidance today (not saying you should've known that, but providing context).

> I thought, it makes sense since these constraints should not be omitted in any cases, and thus be part of the store. WDYT?

It's a fine line to walk and I don't think there's any hard guidelines on what goes into (1) database constraint (2) store method (3) service method.

If you feel that it's always part of `Create` and it's easy to test (sometimes 3 separate store methods make writing tests easier, because you can pick & choose when building up test data in other tests), then that's fine too. Again, not blocking on this, but I just found that "keep store methods small and dumb; compose store methods in service layer" has been a good idea in the past. But devil's in the details :)

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+func (s *accessRequestStore) GetByID(ctx context.Context, id int32) (*types.AccessRequest, error) {
+	row := s.QueryRow(ctx, sqlf.Sprintf("SELECT id, created_at, updated_at, name, email, status, additional_info FROM access_requests WHERE id = %s", id))
```
How about defining all the columns separately? See this for example: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/teams.go?L543:9&subtree=true

Or here: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/webhooks.go?L124:12&subtree=true

mrnugget commented on schema/site.schema.json:
```diff
@@ -488,6 +488,17 @@
+        },
+        "accessRequests": {
+          "description": "Allow users to request access to the instance if signup is disabled",
+          "type": "object",
```
Why not make this a boolean? `enableAccessRequests`?

erzhtor replied:
For this initial version, we probably will keep it simple, but in the future, we might add some extra configuration fields, such as setting allowed domains or smth else, so I thought to have it as an object config already.

mrnugget replied:
In that case, will it still be under `experimentalFeatures`? Or will you move it anyway?

I'm betting on [`YAGNI`](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) :)

erzhtor replied:
I got your point and made it a flat structure (with dot separation `accessRequets.enabled`) instead of a nested object 073a9c5d6c922aa95913c89d184520b6177081f7, so even in case of additional config fields within experimental features, we can add flat but conventionally named field keys.

mrnugget reviewed (commented):
First high-level pass on backend code, haven't looked at frontend code.

Very clean! Left suggestions/asks on making it more match up with our conventions

pjlast commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+		return true
+	}
+
+	// test whether
```
Leftover comment?

pjlast commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,121 @@
+
+// handleRequestAccess handles submission of the request access form.
+func handleRequestAccess(logger log.Logger, db database.DB, w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, fmt.Sprintf("unsupported method %s", r.Method), http.StatusBadRequest)
+		return
+	}
```
Method guarding should be handled by the router

erzhtor replied:
Thanks, updated. P.S: BTW, I followed the [same code guideline](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@erzhtor/add-request-access-feature/-/blob/cmd/frontend/internal/auth/userpasswd/handlers.go?L375) as other public handlers.

pjlast replied:
Ah I see. I'm wondering if all of those are superfluous 🤔 I see I did the same at some point. But my thinking now is that you can't reach the handler without going through the router, and the router has a request guard applied to each route

pjlast commented on schema/site.schema.json:
```diff
@@ -488,6 +488,17 @@
+            "enabled": {
+              "description": "Whether to enable this feature or not",
+              "type": "boolean",
+              "default": true
```
Just be aware that setting a default value in the Schema does not give it a default value in the generated Go code. It will default to `false` since `false` is the default value for Go booleans.

erzhtor replied:
Thanks that is a good point. I've made it a pointer, following other config conventions, e.g. `bacthChanges.enabled`, https://github.com/sourcegraph/sourcegraph/pull/47741#discussion_r1112319931

evict commented on schema/site.schema.json:
```diff
@@ -488,6 +488,17 @@
             "pointer": true
           }
+        },
+        "accessRequests": {
```
Would it be possible for site-admins to configure an email domain that is allowlisted for access-requests, or maybe a list? That way they can make sure the access requests remain in the org?

erzhtor replied:
Thank you for the feedback. It was also proposed the [initial designs](https://www.figma.com/file/XnMKHZ7RpnTU5laBFOZQDg/Request-Flow?node-id=2232%3A1362&t=ANg2fbOnBSOhKWCh-4), but we decided to move it outside of MVP scope for now. I will follow-up after this PR merge. Let me know if you think this is critical and should be included in the current PR. cc @rrhyne @danieldides

erzhtor commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	accessRequest.Status = args.Status
+	if _, err := r.db.AccessRequests().Update(ctx, &types.AccessRequest{ID: accessRequest.ID, Status: accessRequest.Status}); err != nil {
```
Thank you, updated 96afefc242e004b717a514fe5214db317ad9e52e.

erzhtor commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,112 @@
+		return nil, err
+	}
+
+	return &accessRequestsResolver{r.db, args}, nil
```
TIL, Thank you! I suppose you meant [connection_resolver helper](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/graphqlutil/connection_resolver.go). I updated to use it and existing [PageSwitcher.tsx](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/wildcard/src/components/PageSwitcher/PageSwitcher.tsx?subtree=true) component on the frontend.

erzhtor commented:
@sourcegraph/iam, @rrhyne, @taylorsperry, @meglynst, I've addressed PR comments that were previously left and would appreciate it if you could take a look again. I want to merge this PR soon as this one blocks two more ping/analytics-related PRs (https://github.com/sourcegraph/sourcegraph/pull/48026 and https://github.com/sourcegraph/analytics/pull/804).

mrnugget commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,148 @@
+		return nil, err
+	}
+
+	store, err := r.db.AccessRequests().Transact(ctx)
```
Nitpick, but this is usually what we do and it helps spot transactions on first read :)

```suggestion
	tx, err := r.db.AccessRequests().Transact(ctx)
```

mrnugget commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,148 @@
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = store.Done(err) }()
```
Check out this thread from yesterday: https://sourcegraph.slack.com/archives/C3B3SDBMY/p1677103081757369

I'm wondering now whether you don't need to make `err error` the 2nd return value of this method.

sashaostrikov replied:
we need a named error in this case, yep

erzhtor replied:
TIL: Thank you, updated to use a new WithTransact approach 6640c7d5c97dd09706fd3e799473e82763a21873.

mrnugget commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,148 @@
+	if _, err := store.Update(ctx, &types.AccessRequest{ID: accessRequest.ID, Status: accessRequest.Status}); err != nil {
+		return nil, err
+	}
+	return nil, nil
```
```suggestion
	return &EmptyResponse{}, nil
```

mrnugget commented on cmd/frontend/graphqlbackend/node.go:
```diff
@@ -348,3 +348,8 @@ func (r *NodeResolver) ToPermission() (PermissionResolver, bool) {
 	return n, ok
 }
+
+func (r *NodeResolver) ToAccessRequest() (*accessRequestResolver, bool) {
+	n, ok := r.Node.(*accessRequestResolver)
+	return n, ok
+}
```
If it's a node, you also need to make sure you make it queryable by ID. For that you need to add it to this: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/graphqlbackend.go?L611:12

erzhtor replied:
Good catch, added 6640c7d5c97dd09706fd3e799473e82763a21873.

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,276 @@
+
+const (
+	accessRequestInsertQuery = `
+		INSERT INTO access_requests (%s)
+		VALUES ( %s, %s, %s, %s )
+		RETURNING %s`
+	accessRequestListQuery = `
+		SELECT %s
+		FROM access_requests
+		WHERE (%s)`
+	accessRequestUpdateQuery = `
+		UPDATE access_requests
+		SET status = %s
+		WHERE id = %s
+		RETURNING %s`
+)
+
+type AccessRequestListColumn string
+
+const (
+	AccessRequestListID AccessRequestListColumn = "id"
+)
+
+var (
+	accessRequestColumns = []*sqlf.Query{
+		sqlf.Sprintf("id"),
+		sqlf.Sprintf("created_at"),
+		sqlf.Sprintf("updated_at"),
+		sqlf.Sprintf("name"),
+		sqlf.Sprintf("email"),
+		sqlf.Sprintf("status"),
+		sqlf.Sprintf("additional_info"),
+	}
+	accessRequestInsertColumns = []*sqlf.Query{
+		sqlf.Sprintf("name"),
+		sqlf.Sprintf("email"),
+		sqlf.Sprintf("additional_info"),
+		sqlf.Sprintf("status"),
+	}
+)
```
This makes me happy ❤️

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = tx.Done(err) }()
```
Same as above. I wonder whether we don't need a named return parameter here

erzhtor replied:
https://github.com/sourcegraph/sourcegraph/pull/47741#discussion_r1116723724

mrnugget commented on internal/database/access_requests.go:
```diff
@@ -0,0 +1,276 @@
+	return &accessRequest, nil
+}
+
+var scanAccessRequests = basestore.NewSliceScanner(scanAccessRequest)
```
Good job man!

mrnugget reviewed (approved):
Great work!

kopancek commented on client/web/src/site-admin/UserManagement/components/DropdownPagination.tsx:
```diff
@@ -0,0 +1,115 @@
+    return (
+        <div className={classNames('d-flex justify-content-between align-items-center', className)}>
+            <Button className="mr-2" onClick={handlePreviousPage}>
+                <Icon aria-label="Show previous icon" svgPath={mdiChevronLeft} />
```
Since we do not have a label for the button, it might be enough to just say what the button does here instead of saying the `icon` word.
```suggestion
                <Icon aria-label="Show previous" svgPath={mdiChevronLeft} />
```

kopancek commented on client/web/src/site-admin/analytics/components/AnalyticsPageTitle.tsx:
```diff
@@ -1,25 +1,12 @@
-        </Heading>
-    </div>
+    <SiteAdminPageTitle icon={{ svgPath: mdiChartLineVariant, ariaLabel: 'Analytics icon' }}>
+        <span>Analytics</span>
```
Is the span needed here?
```suggestion
        Analytics
```

kopancek commented on client/web/src/site-admin/components/SiteAdminPageTitle.tsx:
```diff
@@ -0,0 +1,36 @@
+    return (
+        <div className="d-flex flex-column justify-content-between align-items-start">
+            <Heading as="h3" styleAs="h2" className="mb-4 mt-2 d-flex align-items-center">
+                <Icon
+                    className="mr-1"
+                    color="var(--link-color)"
+                    svgPath={icon.svgPath}
+                    size="sm"
+                    aria-label={icon.ariaLabel}
+                />
```
Since the icon is not needed, instead of aria-label, provide aria-hidden.

erzhtor replied:
Thank you for the suggestion! Updated!

kopancek commented on client/web/src/util/checkIsRequestAccessAllowed.ts:
```diff
@@ -0,0 +1,15 @@
+ * @returns whether the access request feature should be enabled/allowed
+ */
+export function checkIsRequestAccessAllowed(
+    isSourcegraphDotCom: boolean,
+    allowSignup: boolean,
+    accessRequestsFeatureEnabled: SourcegraphContext['experimentalFeatures']['accessRequests.enabled']
```
All of these things are part of `window.context`. Is it really needed to send them as parameters to this function?

erzhtor replied:
I just thought to avoid making impute utility function by using global variables. Let me know if you think that we should not pass these args and instead use `window.context` for the utility function.

kopancek commented on cmd/frontend/graphqlbackend/access_requests.go:
```diff
@@ -0,0 +1,148 @@
+	database.AccessRequestsFilterArgs
+	graphqlutil.ConnectionResolverArgs
+}) (*graphqlutil.ConnectionResolver[*accessRequestResolver], error) {
+	// 🚨 SECURITY: Only site admins can see access requests.
```
Would it make sense for the user to see their own access request?

erzhtor replied:
These are requests from unauthenticated users, so they don't have an account. P.S.: The idea of "request access" for v1 is to allow users without an account to request an account creation.

kopancek commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,91 @@
+		if builtInAuthProvider != nil && builtInAuthProvider.AllowSignup {
+			logger.Error("signup is enabled, but received access request")
+			http.Error(w, "Use sign up instead.", http.StatusConflict)
+			return
```
In this case, should it redirect to the signup page instead of just returning error?

erzhtor replied:
This is a POST request handler (a form submissions handler). Implicitly redirecting when would be confusing, so used instead an error message.

kopancek commented on client/web/src/site-admin/AccessRequestsPage/index.tsx:
```diff
@@ -0,0 +1,251 @@
+                    />
+                </div>
+                <Grid columnCount={5}>
+                    {['Email', 'Name', 'Last requested at', 'Extra Details', ''].map((value, index) => (
```
`Last requested at` seems misleading, since it is showing `Created at` instead. Should we change it? Or change DB layer to update the date when the user tries to create an access request even if they did already beforehand. That would represent last requested at better.
```suggestion
                    {['Email', 'Name', 'Created at', 'Extra Details', ''].map((value, index) => (
```

erzhtor replied:
Thank you, updated!

kopancek commented on client/web/src/auth/RequestAccessPage.tsx:
```diff
@@ -0,0 +1,210 @@
+                        path="done"
+                        element={
+                            <Text data-testid="request-access-post-submit">
+                                Thank you! We notified the admin of this instance of your request.
```
```suggestion
                                Thank you! We notified the admin of your request.
```

kopancek reviewed (commented):
Looks good in general. Added comments for improvements or where I didn't understand some changes.

Since the PR is quite big, it's also a fair amount of comments..

sashaostrikov commented on cmd/frontend/internal/auth/accessrequest/handlers.go:
```diff
@@ -0,0 +1,91 @@
+	"net/http"
+
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/internal/auth/userpasswd"
+	sgactor "github.com/sourcegraph/sourcegraph/internal/actor"
```
looks like an alias for actor can be skipped here
```suggestion
	"github.com/sourcegraph/sourcegraph/internal/actor"
```

erzhtor replied:
Thank you, updated!

sashaostrikov commented on internal/conf/computed.go:
```diff
@@ -296,6 +296,12 @@ func IsBuiltinSignupAllowed() bool {
+// IsAccessRequestsEnabled returns whether request access experimental feature is enabled or not.
+func IsAccessRequestsEnabled() bool {
+	experimentalFeatures := Get().ExperimentalFeatures
+	return experimentalFeatures == nil || experimentalFeatures.AccessRequestsEnabled == nil || *experimentalFeatures.AccessRequestsEnabled
```
Shouldn't it be other way around?
```suggestion
	return experimentalFeatures != nil && experimentalFeatures.AccessRequestsEnabled != nil && *experimentalFeatures.AccessRequestsEnabled
```

erzhtor replied:
I intentionally used this approach because we want an experimental feature enabled by default, even if not set.

sashaostrikov reviewed (approved):
Took a look mostly at backend part.
Amazing job! 👍🏻 
Left a couple of comments.

meglynst commented:
@erzhtor @rrhyne One question I have in regards to the "Site admin / Access Requests / Approve with error" (the last image listed above). This error language does not match what we have in the [Figma file](https://www.figma.com/file/GQEBnjna5AS5l6vtm4jejD/Account-requests-UI?node-id=140%3A9755&t=OwxWKh1MyIJe6fyZ-0). @rrhyne Do we have a final language update for this error message?

mrnugget commented on client/web/src/auth/SignInPage.tsx:
```diff
@@ -138,6 +149,11 @@ export const SignInPage: React.FunctionComponent<React.PropsWithChildren<SignInP
                     </Text>
+                ) : isRequestAccessAllowed ? (
+                    <Text className="text-muted">
+                        Need an account? <Link to="/request-access">Request access to the admin</Link> or contact your
```
➕ 1 on this. The "Request access to the admin" is wrong and I think the shorter version Milan suggests is perfectly fine.

erzhtor replied:
Thank you, updated!

erzhtor commented on client/web/src/Layout.tsx:
```diff
@@ -113,11 +113,13 @@ export const Layout: React.FC<LegacyLayoutProps> = props => {
+        PageRoutes.PasswordReset,
+        PageRoutes.Welcome,
+        PageRoutes.RequestAccess,
+    ].includes(routeMatch)
```
The PR branch was outdated; after actualizing with the main seems that we already use `useMatches`.

erzhtor commented on client/web/src/auth/RequestAccessPage.tsx:
```diff
@@ -0,0 +1,210 @@
+        setLoading(true)
+        onError(undefined)
+        try {
+            const response = await fetch('/-/request-access', {
```
Because GQL is not available for unauthenticated users, and this particular page and REST API are used by unauthed users to request access (aka account creation)

erzhtor commented on client/web/src/site-admin/UserManagement/components/DropdownPagination.story.tsx:
```diff
@@ -0,0 +1,32 @@
+    return (
+        <DropdownPagination
+            limit={limit}
+            offset={offset}
```
Thank you! This was extracted from the existing UserManagement page to reuse in the new AccessRequestsPage before I learned about PageSwitcher component and generic connection resolver. Reverted extraction back and used existing components.

erzhtor commented on client/web/src/auth/RequestAccessPage.tsx:
```diff
@@ -0,0 +1,210 @@
+
+    if (!isRequestAccessAllowed) {
+        return <Navigate to={PageRoutes.SignIn} replace={true} />
+    }
```
In this scenario when the component auto-redirects based on conditions/props `<Navigate` fits better because of its declarative nature. Whereas `useNavigate` fits well in action handling scenarios, eg. onClick, onSubmit, etc. Let me know please if you think otherwise.

erzhtor commented on client/web/src/site-admin/AccessRequestsPage/index.tsx:
```diff
@@ -0,0 +1,251 @@
+                        </Text>
+                    ))}
+                    {connection?.nodes.map(({ id, email, name, createdAt, additionalInfo }) => (
+                        <Fragment key={email}>
+                            <Text className="mb-0 d-flex align-items-center">{email}</Text>
+                            <Text className="mb-0 d-flex align-items-center">{name}</Text>
+                            <Text className="mb-0 d-flex align-items-center">
```
I'm using [Grid](https://storybook.sgdev.org/?path=/story/wildcard-grid--grid-examples) component, which works only on direct child elements.

erzhtor commented on client/web/src/site-admin/AccessRequestsPage/index.tsx:
```diff
@@ -0,0 +1,251 @@
+ * @param randomize whether to add a random suffix to the username to avoid collisions
+ * @returns username
+ */
+function toUsername(name: string, randomize?: boolean): string {
+    // Remove all non-alphanumeric characters from the name and convert to lowercase.
+    const username = name.replace(/[^\dA-Za-z]/g, '').toLowerCase()
+    if (!randomize) {
+        return username
+    }
+    // Add a random 5-character suffix to the username to avoid collisions.
+    return username + '-' + Math.random().toString(36).slice(2, 7)
+}
+
+/**
+ * A react hook that returns a function that generates a username for a user with the given name.
+ * It checks if the username already exists and if so, it adds a random suffix to the username.
+ */
+function useGenerateUsername(): (name: string) => Promise<string> {
+    const [doesUsernameExist] = useLazyQuery<DoesUsernameExistResult, DoesUsernameExistVariables>(
+        DOES_USERNAME_EXIST,
+        {}
+    )
+
+    return useCallback(
+        async (name: string) => {
+            let username = toUsername(name)
+            while (
+                await doesUsernameExist({
+                    variables: {
+                        username,
+                    },
+                }).then(({ data }) => !!data?.user)
+            ) {
+                username = toUsername(name, true)
+            }
+            return username
+        },
+        [doesUsernameExist]
+    )
+}
```
Currently, I'm using the existing `createUser` mutation/resolver, which is also used in the Users / New page. I want to avoid duplicating code, and extracting that code and reusing it in both the "approve access request" page and the "create new user" page requires some extra effort and writing tests. 

However, thank you for the suggestion. I've added a follow-up https://github.com/sourcegraph/sourcegraph/issues/48244 to tackle this afterward.

erzhtor commented:
> @erzhtor @rrhyne One question I have in regards to the "Site admin / Access Requests / Approve with error" (the last image listed above). This error language does not match what we have in the [Figma file](https://www.figma.com/file/GQEBnjna5AS5l6vtm4jejD/Account-requests-UI?node-id=140%3A9755&t=OwxWKh1MyIJe6fyZ-0). @rrhyne Do we have a final language update for this error message?

@meglynst, @rrhyne, thank you for the review! I've addressed some of the design feedback within this PR. For the left things, I've created follow-up issues (https://github.com/sourcegraph/sourcegraph/issues/48252, https://github.com/sourcegraph/sourcegraph/issues/48251) to tackle them separately, as this PR already grew quite large.

erzhtor commented on client/web/src/LegacyLayout.tsx:
```diff
@@ -148,10 +148,14 @@ export const LegacyLayout: React.FunctionComponent<React.PropsWithChildren<Legac
-        location.pathname === PageRoutes.SignUp ||
-        location.pathname === PageRoutes.PasswordReset ||
-        location.pathname === PageRoutes.Welcome
+        routeMatch &&
```
I clarified from @philipp-spiess as he is working on router migration, and we agreed to leave this LegacyLayout as is to avoid unnecessary regressions.
